### PR TITLE
nsq_weight : skip NSQ partitions based on H vs V costs

### PR DESCRIPTION
**Pull request by chkngit**
## Description
 nsq_hv_level :
        0:  OFF
        1: ON  if H>V + TH1% then  skip HA/HB/H4  
                       if V>H+ TH1% then   skip VA/VB/V4  
      
## Testing and Performance
 in context of cs2 branch / enc-mode 0. 

| Speed Deviation| BD-Rate Deviation|
| -- | --
| ~+1.95% | -0.02%|

### CODE REVIEW

File: Source/Lib/Encoder/Codec/EbProductCodingLoop.c [hunk 1]
```diff
@@ -7498,6 +7503,21 @@ uint8_t update_skip_nsq_shapes(SequenceControlSet *scs_ptr, PictureControlSet *p
...
+            if (!skip_nsq && context_ptr->nsq_hv_level > 0) {
+                if (local_cu_unit[sqi + 3].avail_blk_flag && local_cu_unit[sqi + 4].avail_blk_flag) {
+                    uint64_t v_cost = local_cu_unit[sqi + 3].default_cost + local_cu_unit[sqi + 4].default_cost;
+                    uint32_t offset = 10;
+                    if (context_ptr->nsq_hv_level == 2 && context_ptr->blk_geom->shape == PART_H4)
+                        offset = 5;
+                    if (offset >= 5 && scs_ptr->static_config.qp <= 20)
+                        offset -= 5;
+                    uint32_t v_weight = 100 + offset;
+                    skip_nsq = (h_cost > ((v_cost * v_weight) / 100));
```
Comment (joelsole) [inline Source/Lib/Encoder/Codec/EbProductCodingLoop.c]: There is some magic happening here. Could some clarification on these numbers be added here, besides the explanation already present in EbEncDecProcess.c?

File: Source/Lib/Encoder/Codec/EbProductCodingLoop.c [hunk 2]
```diff
@@ -7536,6 +7556,22 @@ uint8_t update_skip_nsq_shapes(SequenceControlSet *scs_ptr, PictureControlSet *p
...
+
+                    if (context_ptr->nsq_hv_level == 2 && context_ptr->blk_geom->shape == PART_V4)
+                        offset = 5;
+                    if (offset >= 5 && scs_ptr->static_config.qp <= 20)
+                        offset -= 5;
+
+                    uint32_t h_weight = 100 + offset;
+                    skip_nsq = (v_cost > ((h_cost * h_weight) / 100));
+                }
+            }
```
Comment (joelsole) [inline Source/Lib/Encoder/Codec/EbProductCodingLoop.c]: This piece of code is almost the same as the one above, using the same tuning parameters. Could you abstract it out to a common function call?

File: Source/Lib/Encoder/Codec/EbProductCodingLoop.c (hunk 1 above)
Comment (chkngit) [reply Source/Lib/Encoder/Codec/EbProductCodingLoop.c]: //if the cost of H partition is bigger than the V partition by a certain percentage, skip HA/HB
//use 10% to skip HA/HB, use 5% to skip H4, also for very low QP be more aggressive to skip

File: Source/Lib/Encoder/Codec/EbProductCodingLoop.c (hunk 2 above)
Comment (chkngit) [reply Source/Lib/Encoder/Codec/EbProductCodingLoop.c]: we can but the compromise would be to add too much conditional code to streat the vertical and horizontal cases separately, which might make the code look complex to read, I really think this better to stay like that, we can even have some different tuning in the future for the H/V case

File: Source/Lib/Encoder/Codec/EbEncDecProcess.c [hunk 3]
```diff
@@ -1917,6 +1917,22 @@ EbErrorType signal_derivation_enc_dec_kernel_oq(SequenceControlSet * scs_ptr,
         context_ptr->sq_weight = scs_ptr->static_config.sq_weight;
 #endif
 
+#if NSQ_HV
+    // nsq_hv_level  needs sq_weight to be ON
+        // 0: OFF
+        // 1: ON 10% + skip HA/HB/H4  or skip VA/VB/V4
+        // 2: ON 10% + skip HA/HB  or skip VA/VB   ,  5% + skip H4  or skip V4
```
Comment (fouedb) [inline Source/Lib/Encoder/Codec/EbEncDecProcess.c]: Clarify these comments.

File: Source/Lib/Encoder/Codec/EbProductCodingLoop.c (hunk 1 above)
Comment (fouedb) [reply Source/Lib/Encoder/Codec/EbProductCodingLoop.c]: Add in the code the comments from the previous reply.

Comment (chkngit) [reply Source/Lib/Encoder/Codec/EbProductCodingLoop.c]: done

File: Source/Lib/Encoder/Codec/EbEncDecProcess.c (hunk 3 above)
Comment (chkngit) [reply Source/Lib/Encoder/Codec/EbEncDecProcess.c]: done

**Comment by OmranAbdelkafi** [comment]
## Performance Evaluation 

8bit- objective-1-fast list

Speed   Deviation | BD-Rate   Deviation
-- | --
2.56% | 0.06%

**Comment by hassount** [comment]
@chkngit can you please rebase this PR?

**Comment by chkngit** [comment]
@hassount  done rebasing

**Comment by hassount** [comment]
@fouedb is this good to go ?
@chkngit sorry but one more conflict.

**Comment by fouedb** [comment]
> @fouedb is this good to go ?
> @chkngit sorry but one more conflict.

Yes.

**Comment by chkngit** [comment]
@hassount  rebased again

**Comment by hassount** [comment]
> @hassount rebased again

Thank you, will check the tests and merge

**Pull request merged**